### ISSUE TITLE
KG to updated with every single commit

### ISSUE THREAD

**Issue by jachro** [SwissDataScienceCenter/renku#585]
As a user, I'd like to see the Knowledge Graph of the project I'm working with information from all the commits, not just the ones which are pushed to a remote repository.

There are probably several ways of achieving that but the idea we came into is to add a post-commit git hook which would be pushing every single commit user does into some temporary branch.
Pros:
- the solution will work for both remote in-browser Jupyter sessions as well as users working on their local computers;
- no additional services/endpoints have to be created in the Graph-Services;
- no additional security issues as there are no new services or resources involved in what's being used at the moment.
Cons:
- increased communication between user workplace (either Jupyter session or local machine) and GitLab which could eventually make the user experience a bit worse; this could be maybe addressed by doing the post-commit push operation as a non-blocking background operation;
- a minor: some temporary remote branches would have to be created so a user would be able to see them with `git branch -a`.

Things to agree on:
- who and when should create and remove the temporary remote branches; the options seem to be:
  - the git post-commit hook simply does `git push origin renku/pre-push-temp/<unique-id>` and the webhook-service once received and processed a relevant push event from GitLab, removes the branch, so a new branch would be created with a next commit;
    - pros: 
      - a clear picture of who creates a temp branch and who and when removes it;
      - no problem with temp branches removal for cases of users working on their local machines;
    - cons: 
      - the webhook-service would have to remove branches and simply act differently for commits from different branches; it looks it's not really the responsibility of that service to remove branches;
  - a new temporary `renku/pre-push-temp/<unique-id>` branch is created when a user session is created and removed on session termination;
    - pros: 
      - a single place where the temp branches are managed;
      - no changes in the webhook-service;
    - cons: 
      - how/when the temp branches would be deleted for users working on their local machines;
      - how that would work for scenarios like this: a user starts a session -> he creates a branch -> does a commit in the branch -> switches back to master -> creates another branch -> does another commit; I think pushing to a single temp session lived branch could be a bit tricky (a force push maybe?);
- in general, regardless of the mentioned options, should there be another service somewhere which would periodically check for stale temp branches and removing them? how the service would know if that's a stale temp branch and how it even would know where to look for such branches?

**Comment by rokroskar** [comment]
One other potential con of this approach - the history fetched from gitlab will not match the KG. Maybe this would be fine if we marked the unpushed edges/nodes in some way. Or maybe it doesn't really matter...?

**Comment by jachro** [comment]
Good question! I'd be opting for that it does not matter, at least for now.

Actually, there's another question. If a user does some commits but didn't push them at all and then decides to remove all the commit(s) completely, how we'd remove the triples from the graph? Should there be maybe an event sent to remove the triples?

**Comment by rokroskar** [comment]
I suppose this would be the reason for marking these nodes as "tentative" or use some similar terminology. But if the user decides to never push those commits and just keep them in some local branch, you'll never get a delete event of any  kind. Another issue is that if the user commits but then rebases those commits and pushes the rebased version, you will receive the same information but with different commit hashes. Without comparing the content you won't know that they are actually the same thing. Though I suppose the commit hook itself could handle the rebase potentially deleting the affected triples.

**Comment by jachro** [comment]
Yes, it looks then there could be a few corner cases which are not necessarily easy to address...

**Comment by rokroskar** [comment]
there is a [post-rewrite](https://git-scm.com/docs/githooks#_post_rewrite) hook that could be used there...
